src/gt4py/next/embedded/nd_array_field.py

```diff
@@ -174,12 +189,57 @@ def from_array(
     def remap(self: _BaseNdArrayField, connectivity) -> _BaseNdArrayField:
         raise NotImplementedError()
 
+    def __getitem__(self, index: common.FieldSlice) -> common.Field | core_defs.ScalarT:
+        index = _tuplize_field_slice(index)
+
+        if common.is_domain_slice(index):
+            return self._getitem_absolute_slice(index)
+
+        assert isinstance(index, tuple)
+        if all(
+            isinstance(idx, slice) or common.is_int_index(idx) or idx is Ellipsis for idx in index
+        ):
+            return self._getitem_relative_slice(index)
+
+        raise IndexError(f"Unsupported index type: {index}")
+
+    restrict = (
+        __getitem__  # type:ignore[assignment] # TODO(havogt) I don't see the problem that mypy has
+    )
+
+    def __setitem__(
+        self,
+        index: common.FieldSlice,
+        value: common.Field | core_defs.NDArrayObject | core_defs.ScalarT,
+    ) -> None:
```

Comment:
This requires some testing and refactoring (to not duplicate the logic of getitem that much). I'd prefer to do it in a separate PR.